chapters/introduction.tex

```diff
@@ -22,6 +22,11 @@ \section{Scope of the Specification}\label{scope-of-the-specification}
 verified for those classes, before using them to build the simulation
 model.
 
+The restrictions for a simulation model apply when simulating the model.
+Thus a tool may perform parts of the translation of a model lacking some key parts -- provided they are given when simulating the model.
+If the tool has some special mechanism for adjusting the model (like setting parameters), it is only necessary that requirements for a simulation model are satisfied for this adjusted model when simulating it.
```

Comment:
This makes it sound as if this was only a concern for the tool that chooses to delay providing parameter values, but it needs to be formulated to make it clear that any tool needs to be ready to handle models where some parameters aren't given values in the model.

To me it seems that a clarification of the simulation requirement _might_ have a value on its own, but that the case of providing values for parameters is special enough to deserve separate treatment in the context where the requirement for parameters is presented.

For example, I think that until we have the formalization of _structural parameter_, it needs to be formulated as a quality of implementation to which degree "unbound parameters" will work depending on which parameters a tool needs to evaluate during translation

I also think it would be good to somehow emphasize that just using the `start` (aka "the type's default") for any unbound parameter, knowing that it _can_ be given a value after translation, is not good enough.  The tool really needs to represent that the parameter is lacking a value so that it can be verified that one has been given explicitly before simulation starts.

Reply:
> This makes it sound as if this was only a concern for the tool that chooses to delay providing parameter values, but it needs to be formulated to make it clear that any tool needs to be ready to handle models where some parameters aren't given values in the model.

No, I don't see that as necessary.

**The concern for the standard is that you want to simulate a model with possibly changed parameter values, and that modified model should satisfy the requirement for a simulation model in the standard.**

Whether a tool allows you to compile the model without parameter values, and later set them, or requires that they are set before compiling the model shouldn't matter from the specification point of view; as the standard don't need to consider these compiled model, but only simulation of models with parameter values.

Note that the decision was that such handling was "Quality of implementation to be able to translate."

Reply:
I needs to be clear though what high versus low quality of implementation means in this case.

It is not going to be good for portability of Modelica models between tools if we end up with a single tool (System Modeler) that considers a model valid even if some if the parameters are not given values.  Therefore, the specification needs to put pressure on all tools to not opt for low quality of implementation.

Reply:
> I needs to be clear though what high versus low quality of implementation means in this case.
> 
> It is not going to be good for portability of Modelica models between tools if we end up with a single tool (System Modeler) that considers a model valid even if some if the parameters are not given values. Therefore, the specification needs to put pressure on all tools to not opt for low quality of implementation.

**This is not the way to think.** The goal is that the specification should say that a tool is superior in some way, but about different ways of implementing the functionality.

The specification should state the requirements on models (e.g., that parameters must have a value to simulate the model).
_How_ that is verified is up to the tools - and the change of the specification proposed here is that parameter changes _may_ be considered that aren't part of the actual Modelica model in some way and also that this can be applied after some translation; not that tools _should_ handle that. 

For "how" we have previously used "Quality of implementation" as a short-hand for this difference in implementation, and not as a mere ranking system for tools. Thus calling something "low quality" is leading us astray here.

Instead there are different ways of handling this leading to different work-flows for users with different trade-offs and for the standard we should focus on having a clear specification, and let tools innovate in finding work-flows for this that work instead of ranking them ahead of time.

Reply:
My only real concern here is portability of models between tools.

Let's say we have `MyLib.Examples.ParameterizeMe` in a library we either can't or don't want to modify:
```
model ParameterizeMe
  parameter Real p;
  Real y = time / p;
  annotation(experiment(StopTime = 10.0));
end ParameterizeMe;
```

My point is that this is about to become a valid simulation model – it currently isn't.  If this example doesn't work across tools (because the user isn't asked to specify a good value for `p`), we will have a problem with the portability between tools that we are working so hard to achieve.

Reply:
It is not about to become a valid simulation model in itself. `ParameterizeMe` **combined with** `p=...` is about to become a valid simulation model. How to actually combine that is tool-specific, because it involves so many concepts that aren't defined the standard.

Reply:
Call it what you want, the point of #2952 is still that tools need to figure out _some_ way to deal with this model.  The problem with the current text is the _the tool has some special mechanism_ part; the `ParameterizeMe` is a good simulation model without being conditioned on the tool at hand.

Since dealing with `ParameterizeMe` is actually non-trivial compared to what is required of tools today, it really deserves special treatment in the context of equations for parameters rather than being mentioned in passage in the document introduction.

If the `ParameterizeMe` situation is handled in [8.6 _Initialization, initial equation, and initial algorithm_](https://specification.modelica.org/master/equations.html#initialization-initial-equation-and-initial-algorithm), then I'm not sure what the remaining part in the introduction would actually refer to, and would prefer not having this kind of generic statement if we even don't know what it would imply.

Reply:
To avoid the specific issue with `ParameterizeMe` consider:
```
model ParameterizeMe2
  parameter Real p(start=2);
  Real y = time / p;
  annotation(experiment(StopTime = 10.0));
end ParameterizeMe2;
```
If you translate this model, it is stated that a warning is recommended for `p`.
However, if you translate and set `p` to some value before simulating this modified model (or perhaps more accurately scenario as it is assumed that modification isn't stored in the model) that warning is unnecessary.

Note that the previous sentence is deliberately vague on whether you set the value before or after translating, as long as it is done before simulating.

The goal was only to give tools the freedom to handle this, not to require that tools must support such extra modifications in some specific way.

Reply:
The point of `ParameterizeMe` is to not use the parameter `start` trick, either because no particular value to use for `start` is be available at the time of creating the model, or in order to ensure that the parameter is given a meaningful value before simulation.

Reply:
Ignore everything about the original `ParameterizeMe` for the moment and look at `ParameterizeMe2`:
```
model ParameterizeMe2
  parameter Real p(start=2);
  Real y = time / p;
  annotation(experiment(StopTime = 10.0));
end ParameterizeMe2;
```
Currently the specification states that a warning is recommend for `p` when translation.

The proposal is that if you in your tool in some way set `p` (without changing the Modelica source code) before simulating that warning can be skipped, since the 'simulation model' requirement are only important for the model with possible modifications that you actually simulate.

Do you agree that it would be a good change?

Reply:
Yes!

Reply:
Based on this, do you still see a problem with the proposed text - and if so how to improve it?

Reply:
I find it too difficult to grasp the consequences of this sort of generic formulation, and it is also not at all the kind of fix for #2952 that was sought.

For me, the particular problem of providing values for parameters at the start of simulation is concrete enough to be discussed, but also technically challenging enough to deserve its own specific statements in [8.6](https://specification.modelica.org/master/equations.html#initialization-initial-equation-and-initial-algorithm), for example right after the `start` attribute rule for parameters.

Once we have an agreement about the specifics of #2952 we can then think about other parts of the specification that could be eased up similarly.  However, as such changes are not unlikely to involve non-trivial implementation effort in tools, I expect each of them to require careful consideration by the language group before getting adopted.

We must not forget that this is a question about portability between tools in the end.  We don't want System Modeler to be the only tool that can handle (and thus be used to produce) models with unbound parameters; we want the models developed with System Modeler to also work in other tools.  At least, we would need other tools to recognize these models as valid Modelica, even if the rest of their implementation of this feature isn't complete.  Then, if I were a user of such a tool, I'd be quite disappointed if it turned out that the tool was no longer able to prevent me from simulating models without having meaningful values set for all parameters – a feature I expect all tools to have today when this is still something that can be verified during translation.